pkg/agent/openflow/network_policy.go

```diff
@@ -598,54 +704,159 @@ func (c *client) InstallPolicyRuleFlows(rule *types.PolicyRule) error {
+	if rule.From != nil {
+		nClause += 1
+		fromID = nClause
+	}
+	if rule.To != nil {
+		nClause += 1
+		toID = nClause
+	}
+	if rule.Service != nil {
+		nClause += 1
+		serviceID = nClause
+	}
 
-	// Install conjunctive match flows if exists in rule.Form/To/Service
 	var defaultTable binding.Table
 	if rule.From != nil {
 		if isEgressRule {
 			defaultTable = dropTable
 		} else {
 			defaultTable = nil
 		}
-		conj.fromClause = conj.newClause(fromID, nClause, ruleTable, defaultTable)
-		if err := conj.fromClause.addAddrFlows(c, types.SrcAddress, rule.From); err != nil {
-			return err
-		}
+		c.fromClause = c.newClause(fromID, nClause, ruleTable, defaultTable)
 	}
 	if rule.To != nil {
 		if !isEgressRule {
 			defaultTable = dropTable
 		} else {
 			defaultTable = nil
 		}
-		conj.toClause = conj.newClause(toID, nClause, ruleTable, defaultTable)
-		if err := conj.toClause.addAddrFlows(c, types.DstAddress, rule.To); err != nil {
-			return err
-		}
+		c.toClause = c.newClause(toID, nClause, ruleTable, defaultTable)
 	}
 	if rule.Service != nil {
-		conj.serviceClause = conj.newClause(serviceID, nClause, ruleTable, nil)
-		if err := conj.serviceClause.addServiceFlows(c, rule.Service); err != nil {
-			return err
-		}
+		c.serviceClause = c.newClause(serviceID, nClause, ruleTable, nil)
 	}
-	c.policyCache.Store(rule.ID, conj)
-	return nil
+	return nClause, ruleTable, dropTable
 }
 
-func (c *policyRuleConjunction) newClause(clauseID uint8, nClause uint8, ruleTable, dropTable binding.Table) *clause {
-	return &clause{
-		ruleTable: ruleTable,
-		dropTable: dropTable,
-		matches:   make(map[string]*conjMatchFlowContext, 0),
-		action: &conjunctiveAction{
-			conjID:   c.id,
-			clauseID: clauseID,
-			nClause:  nClause,
-		},
+// getConjunctiveMatchFlowAddChange returns the conjMatchFlowContextChanges of the new policyRuleConjunction. It
+// will calculate the expected conjMatchFlowContex status, and the changed Openflow entries.
+func (c *policyRuleConjunction) getConjunctiveMatchFlowAddChange(clnt *client, rule *types.PolicyRule) []*conjMatchFlowContextChange {
+	var ctxChanges []*conjMatchFlowContextChange
+	if c.fromClause != nil {
+		ctxChanges = append(ctxChanges, c.fromClause.addAddrFlows(clnt, types.SrcAddress, rule.From)...)
+	}
+	if c.toClause != nil {
+		ctxChanges = append(ctxChanges, c.toClause.addAddrFlows(clnt, types.DstAddress, rule.To)...)
+	}
+	if c.serviceClause != nil {
+		ctxChanges = append(ctxChanges, c.serviceClause.addServiceFlows(clnt, rule.Service)...)
+	}
+	return ctxChanges
+}
+
+// getConjunctiveMatchFlowDeleteChanges returns the conjMatchFlowContextChanges of the deleted policyRuleConjunction. It
+// will calculate the expected conjMatchFlowContex status, and the changed Openflow entries.
+func (c *policyRuleConjunction) getConjunctiveMatchFlowDeleteChanges() []*conjMatchFlowContextChange {
+	getClauseMatchChanges := func(c *clause, ctxChanges []*conjMatchFlowContextChange) []*conjMatchFlowContextChange {
```

Comment:
Consider to be consistent with getConjunctiveMatchFlowAddChange() and so do not use anonymous func.

Reply:
Updated.